services/audit/record.go

```diff
@@ -0,0 +1,511 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package audit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	audit_model "code.gitea.io/gitea/models/audit"
+	auth_model "code.gitea.io/gitea/models/auth"
+	git_model "code.gitea.io/gitea/models/git"
+	organization_model "code.gitea.io/gitea/models/organization"
+	perm_model "code.gitea.io/gitea/models/perm"
+	repository_model "code.gitea.io/gitea/models/repo"
+	secret_model "code.gitea.io/gitea/models/secret"
+	user_model "code.gitea.io/gitea/models/user"
+	webhook_model "code.gitea.io/gitea/models/webhook"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+type Event struct {
+	Action    audit_model.Action `json:"action"`
+	Actor     TypeDescriptor     `json:"actor"`
+	Scope     TypeDescriptor     `json:"scope"`
+	Target    TypeDescriptor     `json:"target"`
+	Message   string             `json:"message"`
+	Time      time.Time          `json:"time"`
+	IPAddress string             `json:"ip_address"`
+}
+
+func buildEvent(ctx context.Context, action audit_model.Action, actor *user_model.User, scope, target any, message string, v ...any) *Event {
+	return &Event{
+		Action:    action,
+		Actor:     typeToDescription(actor),
+		Scope:     scopeToDescription(scope),
+		Target:    typeToDescription(target),
+		Message:   fmt.Sprintf(message, v...),
+		Time:      time.Now(),
+		IPAddress: tryGetIPAddress(ctx),
+	}
+}
+
+func record(ctx context.Context, action audit_model.Action, actor *user_model.User, scope, target any, message string, v ...any) {
+	if !setting.Audit.Enabled {
+		return
+	}
+
+	e := buildEvent(ctx, action, actor, scope, target, message, v...)
+
+	if err := writeToFile(e); err != nil {
+		log.Error("Error writing audit event to file: %v", err)
+	}
+	if err := writeToDatabase(ctx, e); err != nil {
+		log.Error("Error writing audit event to database: %v", err)
+	}
+}
+
+func RecordUserImpersonation(ctx context.Context, impersonator, target *user_model.User) {
+	record(ctx, audit_model.UserImpersonation, impersonator, impersonator, target, "User %s impersonating user %s.", impersonator.Name, target.Name)
```

Comment:
The descriptions are English hardcoded, they should be translated according to the UI languages.

Reply:
No.
That is neither maintainable nor possible.

Reply:
It's possible if it does not store the description in the database but only struct data and when displaying, uses an i18n string to format it.

Reply:
Yes, that's possible.
Still, I don't think it's an urgent request that can still be added afterward.

Reply:
A prototype is in https://github.com/KN4CK3R/gitea/pull/6 It works but I'm afraid that's very unstable. If the translation is updated in the parameter number, we would have to migrate the database values too.